### PR TITLE
New OSparc simulation APIs

### PR DESCRIPTION
**Pull request by alan-wu**
# Description

Add new OSparc simulation apis to work with the new simulation viewer. Various improvement on SciCrunch query. These new APIs can be tested by clicking on Run Simulation [here](https://mapcore-demo.org/staging/sparc-integrated-features/maps/?id=6bfd76a4).

These changes are implemented by @agarny @Tehsurfer and @alan-wu.

Several improvements on the current SciCrunch APIs to support the latest version of map viewer.

## Type of change

Delete those that don't apply.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Unit tests have been ran and passed.
Functionalities have been tested on this [page](https://mapcore-demo.org/staging/sparc-integrated-features/maps). 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added unit tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -29,6 +29,10 @@
 from app.serializer import ContactRequestSchema
 from app.utilities import img_to_base64_str
 
+import app.osparc as osparc
+import requests
```
Comment (hsorby) [inline app/main.py:33]: Already imported.

File: app/main.py [hunk 2]
```diff
@@ -29,6 +29,10 @@
...
+
+# from pymongo import MongoClient
```
Comment (hsorby) [inline app/main.py:35]: Delete.

File: app/main.py [hunk 3]
```diff
@@ -375,6 +379,33 @@ def sci_doi(doi1, doi2):
...
+    data = {
+        "size": 20,
+        "from": 0,
+        "query": {
+            "query_string": {
+                "fields": fields,
+                "query": curie
+            }
+        }
+    }
```
Comment (hsorby) [inline app/main.py:397]: Query construction should be put into the separate query module.

File: app/main.py [hunk 4]
```diff
@@ -552,9 +563,9 @@ def filter_search(query):
         results = process_results(response.json())
     except requests.exceptions.HTTPError as err:
         logging.error(err)
-        return jsonify({'error': str(err), 'message': 'Sci-crunch is not currently reachable, please try again later'}), 502
-    except json.JSONDecodeError:
-        return jsonify({'message': 'Could not parse Sci-crunch output, please try again later',
+        return jsonify({'error': str(err), 'message': 'SciCrunch is not currently reachable, please try again later'}), 502
+    except json.JSONDecodeError as e:
```
Comment (hsorby) [inline app/main.py:567]: `e` from `as e` isn't used so it shouldn't be added.

File: app/scicrunch_processing_common.py [hunk 5]
```diff
@@ -16,6 +16,7 @@
 VERSION = 'version'
 README = 'readme'
 TITLE = 'title'
+PUBLISH_DATE = 'publishDate'
```
Comment (hsorby) [inline app/scicrunch_processing_common.py:19]: If this is intended to get through the processed results it needs to be added to PASS_THROUGH_KEYS.

**Review by hsorby — CHANGES_REQUESTED** [review]

File: app/scicrunch_processing_v_1_1_3.py [hunk 6]
```diff
@@ -146,18 +146,22 @@
...
     'updated': ['dates', 'updated'],
     'organs': ['anatomy', 'organ'],
+    'organisms': ['organisms', 'subject'],
     'contributors': ['contributors'],
     'doi': ['item', 'curie'],
     'files': ['objects'],
     'version': ['item', 'version', 'keyword'],
-    's3uri': ['pennsieve', 'uri']
+    's3uri': ['pennsieve', 'uri'],
+    'publishDate': ['pennsieve', 'firstPublishedAt', 'timestamp']
```
Comment (Tehsurfer) [inline app/scicrunch_processing_v_1_1_3.py:164]: I think publishDate also needs to be added to the pass through keys? I'll send a pr

**Review by Tehsurfer — CHANGES_REQUESTED** [review]

**Review by Tehsurfer — APPROVED** [review]
Just got an update from @alan-wu that the pass through keys are actually not needed for `publishDate`

This branch is working well and the changes look good to me

**Review by Tehsurfer — APPROVED** [review]

File: app/scicrunch_processing_common.py (hunk 5 above)
Comment (Tehsurfer) [reply app/scicrunch_processing_common.py]: Yeah I thought that might be the case

**Review by agarny — APPROVED** [review]

File: app/osparc.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+        # Upload the configuration file.
+
+        files_api = osparc.FilesApi(api_client)
+
+        try:
+            config_file = files_api.upload_file(temp_config_file.name)
+        except:
+            raise Exception(
+                "the simulation configuration file could not be uploaded")
```
Comment (hsorby) [inline app/osparc.py:30]: You should be raising your own `SimulationException` here and catching that below.  Using `Exception` is the wrong thing to do.

File: app/main.py [hunk 8]
```diff
@@ -575,8 +563,8 @@ def get_facets(type_):
         try:
             json_result = response.json()
             results.append(json_result)
-        except json.JSONDecodeError:
-            return jsonify({'message': 'Could not parse Sci-crunch output, please try again later',
+        except BaseException:
```
Comment (hsorby) [inline app/main.py:566]: Why has this changed from `JSONDecodeError`

File: app/osparc.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
+from app.config import Config
+import json
+import osparc
+import tempfile
+from time import sleep
+
+
+def run_simulation(model_url, json_config):
+    temp_config_file = tempfile.NamedTemporaryFile(mode="w+")
```
Comment (hsorby) [inline app/osparc.py:9]: Use `with tempfile.NamedTemporaryFile() as temp:` context

File: app/scicrunch_requests.py [hunk 10]
```diff
@@ -211,22 +211,41 @@ def facet_query_string(query, terms, facets, type_map):
...
-    if query != "":
+    if query:
         qt = f'({query})'
 
-    if query != "" and len(t) > 0:
+    if query and t:
         qt += " AND "
 
     # Add the brackets and OR and AND parameters
     for k in t:
```
Comment (hsorby) [inline app/scicrunch_requests.py:221]: The variable names used throughout this function are non-descriptive.

**Review by hsorby — CHANGES_REQUESTED** [review]

File: app/scicrunch_requests.py [hunk 11]
```diff
@@ -211,10 +211,10 @@ def facet_query_string(query, terms, facets, type_map):
 
     # Add search query if it exists
     qt = ""
-    if query:
+    if query != "":
```
Comment (agarny) [inline app/scicrunch_requests.py]: I believe I made those changes after PyCharm (?) complained to me. Same for the change below. I believe @hsorby was commenting about the name of some of the variables. For instance, `t` doesn't mean anything to me indeed.

File: app/osparc.py (hunk 9 above)
Comment (agarny) [reply app/osparc.py:9]: Done in https://github.com/agarny/sparc-api/commit/74676553fbb07339ea26653f8d2c0ad1508ef846.

File: app/osparc.py (hunk 7 above)
Comment (agarny) [reply app/osparc.py:30]: Done in https://github.com/agarny/sparc-api/commit/4ce5e7e3fce1ba774dcdcaa8420483a51d65ee1c.

FTR, why is using `Exception` directly the wrong thing to do?

File: app/main.py [hunk 12]
```diff
@@ -368,7 +368,7 @@ def sci_doi(doi1, doi2):
 def sci_organ():
     fields = request.args.getlist('field')
     curie = request.args.get('curie')
-    size = request.args.get('size')
+    size = request.args.get('size')git statu
```
Comment (agarny) [inline app/main.py]: Oopsie!

**Comment by agarny** [comment]
@alan-wu, I believe I have addressed @hsorby's comments. I wanted to push them directly in your branch, but couldn't. So, you can find them in the `simulation` branch of my copy of `sparc-api`.

File: app/main.py (hunk 12 above)
Comment (alan-wu) [reply app/main.py]: Ya, oopsie.

File: app/osparc.py (hunk 7 above)
Comment (hsorby) [reply app/osparc.py]: It is too broad every exception in Python is an `Exception`.

Comment (agarny) [reply app/osparc.py]: Fair enough.

**Review by hsorby — APPROVED** [review]
Not a fan of the variable names used in `facet_query_string` but ...

**Pull request merged**